### PR TITLE
 Running XCUI tests on simulators and devices, skip installing WDA 

### PR DESCRIPTION
**Pull request by MaxCiv**
Ability to run XCUI tests by `xcodebuild test-without-building...`, set environment variables by `xcrun simctl spawn $udid launchctl setenv ENV_NAME ENV_VALUE` and skip installing WebDriverAgent by setting desired capability `"use_wda": false`.
____
Endpoint `/devices/${ref}/run_xcuitest` takes following JSON (`timeout_sec` has default value 300):
```
{
    "app_name": "com.my.app",
    "xctestrun_file_name":"test.xctestrun",
    "test_name":"ui-tests/TestSome",
    "path_to_dir_with_xctestrun_file":"/tmp/Build/Product",
    "timeout_sec": 240
}
```
It returns a map with execution results (`command`, `exitCode`, `stdOut` and `stdErr`).
____
Endpoint `/devices/${ref}/set_env` takes following JSON:
```
{
    "ENV_NAME1": "ENV_VALUE1",
    "ENV_NAME2": "ENV_VALUE2",
    "ENV_NAME_": "ENV_VALUE_"
}
```
All specified environment variables will be set on selected simulator (setting envs is not supported by physical devices)

Also you can add node-specific environment variables to NodeConfig:
```
{
  "timeouts": {
    "device": 600
  },
  "nodes": [
    {
      "user": "",
      "host": "localhost",
      "simulator_limit": 6,
      "environment_variables": {
        "ENV_NAME1": "ENV_VALUE1",
        "ENV_NAME2": "ENV_VALUE2"
      }
    }
  ]
}
```
This environment variables will be set on booting or resetting of every simulator on _this node_.
___
For Ruby client added:
* `reset_async()` for RemoteDevice
* configurable HTTP read_timeout for DeviceProvider in RemoteDevice
* support of `set_env` and `run_xcuitest` commands

### CODE REVIEW

File: device-server/src/main/kotlin/com/badoo/automation/deviceserver/DeviceServer.kt [hunk 1]
```diff
@@ -233,6 +233,16 @@ fun Application.module() {
                 get("state") {
                     call.respond(devicesController.getDeviceState(param(call, "ref")))
                 }
+                post("set_env") {
```
Comment (NickAb) [inline device-server/src/main/kotlin/com/badoo/automation/deviceserver/DeviceServer.kt:236]: Let's make it `post("environment")` so it's more in line with other endpoints. POST verb already implies `setting`

File: device-server/src/main/kotlin/com/badoo/automation/deviceserver/ios/simulator/Simulator.kt [hunk 2]
```diff
@@ -637,4 +643,28 @@ class Simulator (
         logger.debug(logMarker, "Uninstalling application $bundleId from Simulator $this")
         remote.execIgnoringErrors(listOf("xcrun", "simctl", "uninstall", udid, bundleId))
     }
+
+    override fun setEnvironmentVariables(envs: Map<String, String>) {
+        logger.debug(logMarker, "Setting environment variables $envs for Simulator $this")
+        envs.keys.forEach {
+            remote.shell("xcrun simctl spawn $udid launchctl setenv $it ${envs[it]}")
```
Comment (NickAb) [inline device-server/src/main/kotlin/com/badoo/automation/deviceserver/ios/simulator/Simulator.kt:650]: I am concerned that envs[it] are not escaped, which will potentially cause a problem if I want to set env variable to a value containing space. For example `TEST=123 123`.
I think you can use `ShellUtils.escape`, but needs to be tested for local and ssh modes.

Also, it looks like `launchctl setenv` supports setting multiple env variables in a single call, this will save a lot on process calls, especially over ssh. See:
```
Usage: launchctl setenv <<key> <value>, ...>
```

File: device-server/src/main/kotlin/com/badoo/automation/deviceserver/data/XcuiTestExecutionConfig.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        val appName: String,
+
+        @JsonProperty("xctestrun_file_name")
+        val xctestrunFileName: String,
+
+        @JsonProperty("test_name")
+        val testName: String,
+
+        @JsonProperty("path_to_dir_with_xctestrun_file")
+        val pathToDirWithXctestrunFile: Path,
```
Comment (NickAb) [inline device-server/src/main/kotlin/com/badoo/automation/deviceserver/data/XcuiTestExecutionConfig.kt:17]: Can we merge this property with `xctestrunFileName`, it does not look like we are using them separately anywhere.

File: device-server/src/main/kotlin/com/badoo/automation/deviceserver/ios/device/Device.kt [hunk 4]
```diff
@@ -366,9 +366,26 @@ class Device(
...
+        logger.debug(logMarker, "Running XCUI test '${xcuiTestExecutionConfig.testName}' of '${xcuiTestExecutionConfig.appName}'" +
+                " using '${xcuiTestExecutionConfig.pathToDirWithXctestrunFile}/${xcuiTestExecutionConfig.xctestrunFileName}' on device $this")
+
+        val command = listOf("xcodebuild", "test-without-building", "-xctestrun",
+                "${xcuiTestExecutionConfig.pathToDirWithXctestrunFile}/${xcuiTestExecutionConfig.xctestrunFileName}",
+                "-destination", "platform=iOS,id=$udid", "-only-testing:${xcuiTestExecutionConfig.testName}")
+
+        val result = remote.execIgnoringErrors(command, timeOutSeconds = xcuiTestExecutionConfig.timeoutSec)
+        return mapOf(
+                "command" to result.cmd.joinToString(" "),
```
Comment (NickAb) [inline device-server/src/main/kotlin/com/badoo/automation/deviceserver/ios/device/Device.kt:379]: Could you please create a class XcuiTestExecutionResult and use it here instead of Map<String, String>

File: device-server/src/main/kotlin/com/badoo/automation/deviceserver/data/XcuiTestExecutionConfig.kt [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package com.badoo.automation.deviceserver.data
+
+import com.fasterxml.jackson.annotation.JsonProperty
+import java.nio.file.Path
+
+data class XcuiTestExecutionConfig(
+        @JsonProperty("app_name")
+        val appName: String,
```
Comment (NickAb) [inline device-server/src/main/kotlin/com/badoo/automation/deviceserver/data/XcuiTestExecutionConfig.kt:8]: Looks like this property is unused. The only usages are for logging and tests.

File: device-server/src/main/kotlin/com/badoo/automation/deviceserver/ios/device/Device.kt [hunk 6]
```diff
@@ -366,9 +366,26 @@ class Device(
         remote.fbsimctl.uninstallApp(udid, bundleId)
     }
 
+    fun runXcuiTest(xcuiTestExecutionConfig: XcuiTestExecutionConfig): Map<String, String> {
+        logger.debug(logMarker, "Running XCUI test '${xcuiTestExecutionConfig.testName}' of '${xcuiTestExecutionConfig.appName}'" +
+                " using '${xcuiTestExecutionConfig.pathToDirWithXctestrunFile}/${xcuiTestExecutionConfig.xctestrunFileName}' on device $this")
+
+        val command = listOf("xcodebuild", "test-without-building", "-xctestrun",
+                "${xcuiTestExecutionConfig.pathToDirWithXctestrunFile}/${xcuiTestExecutionConfig.xctestrunFileName}",
```
Comment (NickAb) [inline device-server/src/main/kotlin/com/badoo/automation/deviceserver/ios/device/Device.kt:374]: Do I understand correctly, that this will work on localhost only, as it assumes that files for testing are available on the node host?
What if I want to run tests on a remote node (ssh), do I need to copy files manually beforehand?

File: device-server/src/main/kotlin/com/badoo/automation/deviceserver/host/HostFactory.kt [hunk 7]
```diff
@@ -43,7 +43,8 @@ class HostFactory(
                 ),
                 simulatorLimit = config.simulatorLimit,
                 concurrentBoots = config.concurrentBoots,
-                wdaRunnerXctest = getWdaRunnerXctest(remote.isLocalhost(), wdaSimulatorBundle, REMOTE_WDA_BUNDLE_ROOT)
+                wdaRunnerXctest = getWdaRunnerXctest(remote.isLocalhost(), wdaSimulatorBundle, REMOTE_WDA_BUNDLE_ROOT),
+                initialEnvironmentVariables = config.environmentVariables
```
Comment (NickAb) [inline device-server/src/main/kotlin/com/badoo/automation/deviceserver/host/HostFactory.kt:47]: It seems that we don't need `initialEnvironmentVariables` property, and it should not be part of the host node configuration.
In most cases, one can expect environment variables to be specific to the app being tested, so it is a client responsibility to set env variables on a simulator before or during tests.

Please remove `initialEnvironmentVariables` completly.

File: device-server/src/main/kotlin/com/badoo/automation/deviceserver/host/SimulatorsNode.kt [hunk 8]
```diff
@@ -87,7 +88,9 @@ class SimulatorsNode(
 
             logger.debug(simLogMarker, "Will create simulator $ref")
 
-            val simulator = simulatorFactory.newSimulator(ref, remote, fbSimctlDevice, ports, deviceSetPath, wdaRunnerXctest, concurrentBoot, desiredCaps.headless, fbSimctlDevice.toString())
+            val simulator = simulatorFactory.newSimulator(ref, remote, fbSimctlDevice, ports, deviceSetPath,
+                    wdaRunnerXctest, concurrentBoot, desiredCaps.headless, desiredCaps.useWda, fbSimctlDevice.toString())
+            simulator.initialEnvironmentVariables.putAll(initialEnvironmentVariables)
```
Comment (NickAb) [inline device-server/src/main/kotlin/com/badoo/automation/deviceserver/host/SimulatorsNode.kt:93]: See my comment about `initialEnvironmentVariables` above. I think it should be removed.

File: device-server/src/test/kotlin/com/badoo/automation/deviceserver/NodeConfigTest.kt [hunk 9]
```diff
@@ -49,16 +53,16 @@ class NodeConfigTest {
         val config = JsonMapper().fromJson<NodeConfig>("{}")
 
         val expected = NodeConfig(
-            type = NodeConfig.NodeType.Simulators,
```
Comment (NickAb) [inline device-server/src/test/kotlin/com/badoo/automation/deviceserver/NodeConfigTest.kt:52]: Could you please revert unnecessary style changes, or put them into a separate commit/PR?

**Review by NickAb — CHANGES_REQUESTED** [review]
Could you please rebase your branch and squash your commits into 2 or 3 commits:

1. Add useWDA property
2. Add set env
3. Add run XCUITest

**Comment by NickAb** [comment]
Could you please rebase your branch and squash your commits into 2 or 3 commits:

1. Add useWDA property
2. Add set env
3. Add run XCUITest

If it's not too hard, it would be great if you just open 3 different pull requests, useWDA, set env, and XCUITest. We can merge set env and useWDA PRs quickly, changes relating to these two features look good and there are no questions.
I have some questions for changes related to XCUITest.

**Comment by MaxCiv** [comment]
Ok, will do.

File: device-server/src/main/kotlin/com/badoo/automation/deviceserver/DeviceServer.kt (hunk 1 above)
Comment (MaxCiv) [reply device-server/src/main/kotlin/com/badoo/automation/deviceserver/DeviceServer.kt:236]: Done.
#34

File: device-server/src/main/kotlin/com/badoo/automation/deviceserver/ios/simulator/Simulator.kt (hunk 2 above)
Comment (MaxCiv) [reply device-server/src/main/kotlin/com/badoo/automation/deviceserver/ios/simulator/Simulator.kt:650]: Done.
#34

I've tried different variants with `execIgnoringErrors()`, but they didn't work locally. Method with `remote.shell()` and `ShellUtils.escape()` on variable value works fine.

File: device-server/src/main/kotlin/com/badoo/automation/deviceserver/host/HostFactory.kt (hunk 7 above)
Comment (MaxCiv) [reply device-server/src/main/kotlin/com/badoo/automation/deviceserver/host/HostFactory.kt:47]: Done.
#34

File: device-server/src/main/kotlin/com/badoo/automation/deviceserver/host/SimulatorsNode.kt (hunk 8 above)
Comment (MaxCiv) [reply device-server/src/main/kotlin/com/badoo/automation/deviceserver/host/SimulatorsNode.kt:93]: Done.
#34

File: device-server/src/main/kotlin/com/badoo/automation/deviceserver/ios/device/Device.kt (hunk 6 above)
Comment (MaxCiv) [reply device-server/src/main/kotlin/com/badoo/automation/deviceserver/ios/device/Device.kt:374]: Yes, this will work only if files for testing are available on the node host. And for now you should manually copy files before run tests.

So, should ios-device-server be able to transfer files for testing to all needed nodes by itself (by rsync, for example)? Or lets leave it for tests-runners or third-party utilities?

File: device-server/src/main/kotlin/com/badoo/automation/deviceserver/data/XcuiTestExecutionConfig.kt (hunk 3 above)
Comment (MaxCiv) [reply device-server/src/main/kotlin/com/badoo/automation/deviceserver/data/XcuiTestExecutionConfig.kt:17]: Done.
#36

File: device-server/src/main/kotlin/com/badoo/automation/deviceserver/ios/device/Device.kt (hunk 4 above)
Comment (MaxCiv) [reply device-server/src/main/kotlin/com/badoo/automation/deviceserver/ios/device/Device.kt:379]: Done.
#36

File: device-server/src/main/kotlin/com/badoo/automation/deviceserver/data/XcuiTestExecutionConfig.kt (hunk 5 above)
Comment (MaxCiv) [reply device-server/src/main/kotlin/com/badoo/automation/deviceserver/data/XcuiTestExecutionConfig.kt:8]: Done.
#36

**Comment by NickAb** [comment]
Split into #33 #34 #35 #36
